snapshot-service/src/lib.rs

```diff
@@ -156,30 +154,37 @@ impl UpdateHandle {
                 voting_power,
             } = entry;
 
-            let mut key = Vec::with_capacity(4 + 32 + voting_group.as_bytes().len());
+            let voting_key_bytes = voting_key.as_ref().as_ref();
 
-            match tag_id {
-                None => key.extend(&self.next_tag_id.to_be_bytes()),
-                Some(ref tag_id_raw) => key.extend(&**tag_id_raw),
+            let mut key = Vec::with_capacity(
+                size_of::<TagId>() + voting_key_bytes.len() + voting_group.as_bytes().len(),
+            );
+
+            match &tag_id {
+                Tag::Existing(tag_id) | Tag::New(tag_id) => key.extend(&**tag_id),
             }
 
-            key.extend(voting_key.as_ref().as_ref());
+            key.extend(voting_key_bytes);
             key.extend(voting_group.as_bytes());
 
             batch.insert(key, &u64::from(voting_power).to_be_bytes());
         }
 
         {
             let tag = tag.to_string();
-            let tag_id = tag_id.clone();
             let tags = self.tags.clone();
             let entries = self.entries.clone();
-            let next_tag_id = self.next_tag_id.to_be_bytes();
+            let seqs = self.seqs.clone();
 
             tokio::task::spawn_blocking(move || {
-                (&tags, &entries).transaction(move |(tags, entries)| {
-                    if let None = tag_id {
-                        tags.insert(tag.as_bytes(), &next_tag_id)?;
+                (&tags, &entries, &seqs).transaction(move |(tags, entries, seqs)| {
+                    if let Tag::New(id) = &tag_id {
+                        tags.insert(tag.as_bytes(), id)?;
+                        seqs.insert(
+                            TAG_SEQ_KEY,
+                            &(TagId::from_be_bytes(id.as_ref().try_into().unwrap()) + 1)
+                                .to_be_bytes(),
+                        )?;
```

Comment:
Where do you remove existing keys in case of tag update?

Reply:
If you are talking about removing the tag entry if the input vector is empty, I'm not doing it. But I'll add it in a bit. It probably makes more sense.

Reply:
I mean removing keys that are not present anymore in the new input (i.e. there's no way of removing keys once they're added). Could this be helpful?